### DISCUSSION TITLE
关于多个应用打洞的疑问

### DISCUSSION THREAD

**Discussion by zzjzz9266a** [MikeWang000000/Natter#127] [Q&A]
作者您好，首先很感谢你写了那么方便好用的内网打洞工具，现在我已将内网群晖的自托管相册应用打洞到公网了，并且体验很好，避免了在只有v4环境下的访问问题。
但是目前有个疑问，可能比较小白，就是如果我有多款应用分别对应不同端口需要暴露到公网，目前来看单个natter程序只支持打一个洞，也就是一次只会有一个端口暴露，这个设计是by design的嘛？未来有计划支持一次打多个洞嘛？或者说在不用反代的情况下是否有办法一次打多个洞呢？
再次感谢！

**Comment by MikeWang000000** [comment]
首先，感谢您对 Natter 的认可！

---

这个是有意做成这样的，因为 Natter 有不少参数可供调整 (详见 `--help`)，它们只针对一个端口。

如果加入多端口的设计，比如 UDP 端口绑定 1234 转发到本机的 4567，使用通知脚本 A；TCP 端口绑定 2345 转发到 192.168.1.100，不使用通知脚本... 会变得很复杂。

另外，还有可能某个端口出现问题，导致整个 Natter（也就是所有端口）挂掉的情况。

所以，Natter v2 废弃了原先 v0.9 中支持多端口的设计，每个进程只维护一个端口。如果需要使用多个端口，启动多个 Natter 就可以。

---

单进程下，支持多端口是可能的。不过 Natter 从省事 + 便于维护的角度考虑，就没有这么做了。
